components/base/kinematicbase/execution.go

```diff
@@ -0,0 +1,471 @@
+		startPose = startPoseInFrame.Pose()
+	}
+
+	// Pre-process all steps into a series of velocities
+	ptgk.inputLock.Lock()
+	ptgk.currentExecutingSteps, err = ptgk.arcStepsFromInputs(inputSteps, startPose)
+	arcSteps := ptgk.currentExecutingSteps
+	ptgk.inputLock.Unlock()
+	if err != nil {
+		return tryStop(err)
+	}
+
+	for i := 0; i < len(arcSteps); i++ {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		step := ptgk.currentExecutingSteps[i]
+		ptgk.inputLock.Lock() // In the case where there's actual contention here, this could cause timing issues; how to solve?
+		ptgk.currentIdx = i
+		ptgk.currentInputs = step.arcSegment.StartConfiguration
+		ptgk.inputLock.Unlock()
+
+		timestep := time.Duration(step.durationSeconds*1000*1000) * time.Microsecond
+
+		ptgk.logger.Debugf("step, i %d", i)
+		ptgk.logger.Debug(
+			step.linVelMMps,
+			step.angVelDegps,
+			step.durationSeconds,
+			step.arcSegment,
+			spatialmath.PoseToProtobuf(step.arcSegment.StartPosition),
+		)
+
+		ptgk.logger.CDebugf(ctx,
+			"setting velocity to linear %v angular %v and running velocity step for %s",
+			step.linVelMMps,
+			step.angVelDegps,
+			timestep,
+		)
+
+		err := ptgk.Base.SetVelocity(
+			ctx,
+			step.linVelMMps,
+			step.angVelDegps,
+			nil,
+		)
+		if err != nil {
+			return tryStop(err)
+		}
+		arcStartTime := time.Now()
+		// Now we are moving. We need to do several things simultaneously:
+		// - move until we think we have finished the arc, then move on to the next step
+		// - update our CurrentInputs tracking where we are through the arc
+		// - Check where we are relative to where we think we are, and tweak velocities accordingly
+		for timeElapsed := inputUpdateStepSeconds; timeElapsed <= step.durationSeconds; timeElapsed += inputUpdateStepSeconds {
+			if ctx.Err() != nil {
+				return ctx.Err()
+			}
+			// Account for 1) timeElapsed being inputUpdateStepSeconds ahead of actual elapsed time, and the fact that the loop takes
+			// nonzero time to run especially when using the localizer.
+			actualTimeElapsed := time.Since(arcStartTime)
+			remainingTimeStep := time.Duration(1000*1000*timeElapsed)*time.Microsecond - actualTimeElapsed
+
+			if remainingTimeStep > 0 {
+				utils.SelectContextOrWait(ctx, remainingTimeStep)
+				if ctx.Err() != nil {
+					return tryStop(ctx.Err())
+				}
+			}
+			distIncVel := step.linVelMMps.Y
+			if distIncVel == 0 {
+				distIncVel = step.angVelDegps.Z
+			}
+			currentInputs := []referenceframe.Input{
+				step.arcSegment.StartConfiguration[0],
+				step.arcSegment.StartConfiguration[1],
+				{step.arcSegment.StartConfiguration[2].Value + math.Abs(distIncVel)*timeElapsed},
+			}
+			ptgk.inputLock.Lock()
+			ptgk.currentInputs = currentInputs
+			ptgk.inputLock.Unlock()
+
+			// If we have a localizer, we are able to attempt to correct to stay on the path.
+			// For now we do not try to correct while in a correction.
+			if ptgk.Localizer != nil {
+				actualPose, err := ptgk.Localizer.CurrentPosition(ctx)
```

Comment:
[opt] move all the code in this conditional into a separate function for ease of readability.

Reply:
It started out that way, but due to the loop control plus the variables edited, it was kind of a huge pain to manage the scope of everything and return a reasonable number of things.

Reply:
I have successfully broken this up.